File: scripts/seed/builders.py
"""
Text builders for each source type.

Mirrors the JS buildXxxText() functions in ingest.html exactly so the
plain-text output is in the same format the embedding pipeline already knows.
"""


def build_jira_text(d: dict) -> str:
    lines = [
        f"JIRA ISSUE: {d['key']}",
        f"Project: {d.get('project', '')}",
        f"Type: {d.get('issue_type', 'Task')}",
        f"Summary: {d['summary']}",
    ]
    meta_parts = []
    if d.get("status"):
        meta_parts.append(f"Status: {d['status']}")
    if d.get("priority"):
        meta_parts.append(f"Priority: {d['priority']}")
    if d.get("reporter"):
        meta_parts.append(f"Reporter: {d['reporter']}")
    if meta_parts:
        lines.append("   ".join(meta_parts))
    if d.get("labels"):
        lines.append(f"Labels: {d['labels']}")
    lines.append("")
    lines.append("Description:")
    lines.append(d.get("description", ""))
    if d.get("comments"):
        lines.append("")
        lines.append("Comments:")
        for c in d["comments"]:
            lines.append(c)
    return "\n".join(lines)


def build_confluence_text(d: dict) -> str:
    meta_parts = []
    if d.get("space"):
        meta_parts.append(f"Space: {d['space']}")
    if d.get("author"):
        meta_parts.append(f"Author: {d['author']}")
    if d.get("status"):
        meta_parts.append(f"Status: {d['status']}")
    if d.get("confidentiality"):
        meta_parts.append(f"Confidentiality: {d['confidentiality']}")
    lines = [
        f"CONFLUENCE PAGE: {d['title']}",
        "   ".join(meta_parts),
    ]
    if d.get("labels"):
        lines.append(f"Labels: {d['labels']}")
    lines.append("")
    lines.append(d.get("content", ""))
    return "\n".join(lines)


def build_slack_text(d: dict) -> str:
    channel_type = d.get("channel_type", "public")
    lines = [
        f"SLACK THREAD: {d['title']}",
        f"Workspace: {d.get('workspace', '')}   Channel: #{d['channel']} ({channel_type})",
    ]
    if d.get("participants"):
        lines.append(f"Participants: {d['participants']}")
    if d.get("labels"):
        lines.append(f"Labels: {d['labels']}")
    if d.get("linked_jira_tickets"):
        lines.append(f"Linked Jira: {d['linked_jira_tickets']}")
    if d.get("linked_github_prs"):
        lines.append(f"Linked PRs: {d['linked_github_prs']}")
    lines.append("")
    lines.append("Messages:")
    for msg in d.get("messages", []):
        lines.append(msg)
    return "\n".join(lines)


def build_hubspot_text(d: dict) -> str:
    meta_parts = []
    if d.get("company_domain"):
        meta_parts.append(f"Domain: {d['company_domain']}")
    if d.get("stage"):
        meta_parts.append(f"Stage: {d['stage']}")
    if d.get("account_tier"):
        meta_parts.append(f"Tier: {d['account_tier']}")
    if d.get("industry"):
        meta_parts.append(f"Industry: {d['industry']}")
    lines = [
        f"HUBSPOT ACCOUNT: {d['company_name']}",
        "   ".join(meta_parts),
    ]
    if d.get("interested_products"):
        lines.append(f"Products: {d['interested_products']}")
    lines.append("")
    lines.append("Notes:")
    lines.append(d.get("notes", ""))
    if d.get("next_step"):
        lines.append("")
        lines.append(f"Next Step: {d['next_step']}")
    if d.get("blockers"):
        lines.append("")
        lines.append("Blockers:")
        for b in d["blockers"]:
            lines.append(b)
    return "\n".join(lines)


def build_github_text(d: dict) -> str:
    item_type = d.get("type", "pull_request").upper()
    meta_parts = []
    if d.get("state"):
        meta_parts.append(f"State: {d['state']}")
    if d.get("author"):
        meta_parts.append(f"Author: {d['author']}")
    if d.get("base_branch"):
        meta_parts.append(f"Base: {d['base_branch']}")
    lines = [
        f"GITHUB {item_type}: {d['repo']}#{d['number']}",
        f"Title: {d['title']}",
        "   ".join(meta_parts),
    ]
    if d.get("labels"):
        lines.append(f"Labels: {d['labels']}")
    if d.get("files_changed"):
        lines.append(f"Files Changed: {d['files_changed']}")
    if d.get("linked_issues"):
        lines.append(f"Linked Issues: {d['linked_issues']}")
    lines.append("")
    lines.append("Description:")
    lines.append(d.get("body", ""))
    if d.get("comments"):
        lines.append("")
        lines.append("Comments:")
        for c in d["comments"]:
            lines.append(c)
    return "\n".join(lines)

File: frontend/src/components/Sources/builders.ts
/** Mirror of ingest.html buildXxxText() functions — produces plain-text documents. */

export function buildJiraText(d: Record<string, string>): string {
  const lines = [
    `JIRA ISSUE: ${d.key}`,
    `Project: ${d.project}`,
    `Type: ${d.issue_type || d.issueType}   Status: ${d.status}   Priority: ${d.priority || "P2"}`,
    d.reporter ? `Reporter: ${d.reporter}` : "",
    "",
    `Summary: ${d.summary}`,
    "",
    "Description:",
    d.description || "",
  ];
  if (d.labels) lines.push("", `Labels: ${d.labels}`);
  if (d.comments) {
    lines.push("", "Comments:");
    d.comments.split("\n").forEach((c) => c.trim() && lines.push(`  ${c.trim()}`));
  }
  return lines.filter((l) => l !== undefined).join("\n").trim();
}

export function buildConfluenceText(d: Record<string, string>): string {
  const lines = [
    `CONFLUENCE PAGE: ${d.title}`,
    `Space: ${d.space}   Author: ${d.author || "unknown"}   Status: ${d.status || "published"}   Confidentiality: ${d.confidentiality || "internal"}`,
    "",
    d.content || "",
  ];
  if (d.labels) lines.push("", `Labels: ${d.labels}`);
  return lines.join("\n").trim();
}

export function buildHubspotText(d: Record<string, string>): string {
  const lines = [
    `HUBSPOT ACCOUNT: ${d.company_name}`,
    `Domain: ${d.company_domain}   Stage: ${d.stage || "lead"}   Tier: ${d.account_tier || "startup"}   Industry: ${d.industry || ""}`,
    "",
    d.notes ? `Notes:\n${d.notes}` : "",
  ];
  if (d.contacts) lines.push("", `Contacts: ${d.contacts}`);
  if (d.open_deals) lines.push(`Open Deals: ${d.open_deals}`);
  if (d.labels) lines.push("", `Labels: ${d.labels}`);
  return lines.join("\n").trim();
}

export function buildSlackText(d: Record<string, string>): string {
  const lines = [
    `SLACK THREAD: ${d.title}`,
    `Workspace: ${d.workspace}   Channel: #${d.channel} (${d.channel_type || "public"})`,
  ];
  if (d.participants) lines.push(`Participants: ${d.participants}`);
  lines.push("", "Messages:");
  (d.messages || "").split("\n").forEach((m) => m.trim() && lines.push(`  ${m.trim()}`));
  if (d.linked_jira_tickets) lines.push("", `Linked Jira: ${d.linked_jira_tickets}`);
  if (d.linked_github_prs) lines.push(`Linked PRs: ${d.linked_github_prs}`);
  if (d.labels) lines.push("", `Labels: ${d.labels}`);
  return lines.join("\n").trim();
}

export function buildGithubText(d: Record<string, string>): string {
  const kind = (d.type || "pull_request").toUpperCase().replace("_", " ");
  const lines = [
    `GITHUB ${kind}: ${d.repo}#${d.number}`,
    `Title: ${d.title}`,
    `State: ${d.state || "open"}   Author: ${d.author || ""}   Base: ${d.base_branch || "main"}`,
    "",
    "Description:",
    d.body || "",
  ];
  if (d.labels) lines.push("", `Labels: ${d.labels}`);
  if (d.files_changed) lines.push("", `Files changed: ${d.files_changed}`);
  if (d.comments) {
    lines.push("", "Comments:");
    d.comments.split("\n").forEach((c) => c.trim() && lines.push(`  ${c.trim()}`));
  }
  return lines.join("\n").trim();
}

export type SourceType = "jira" | "confluence" | "hubspot" | "slack" | "github";

const BUILDERS: Record<SourceType, (d: Record<string, string>) => string> = {
  jira: buildJiraText,
  confluence: buildConfluenceText,
  hubspot: buildHubspotText,
  slack: buildSlackText,
  github: buildGithubText,
};

const FILE_NAMES: Record<SourceType, (d: Record<string, string>) => string> = {
  jira:       (d) => `jira-${d.key || d.summary?.slice(0, 20) || "issue"}.txt`,
  confluence: (d) => `confluence-${d.space}-${d.title?.slice(0, 30) || "page"}.txt`,
  hubspot:    (d) => `hubspot-${d.company_domain || d.company_name || "account"}.txt`,
  slack:      (d) => `slack-${d.channel}-${d.title?.slice(0, 30) || "thread"}.txt`,
  github:     (d) => `github-${(d.repo || "repo").replace("/", "_")}-${d.number || "0"}.txt`,
};

/** Build a File ready for upload from source form data. */
export function buildFile(sourceType: SourceType, data: Record<string, string>): File {
  const text = BUILDERS[sourceType](data);
  const name = FILE_NAMES[sourceType](data);
  return new File([text], name, { type: "text/plain" });
}
